modules/posix/src/ciaaPOSIX_unistd.c

```diff
@@ -0,0 +1,234 @@
+ **
+ **/
+static void ciaaPOSIX_sleepAlgorithm(uint32_t toSleep);
+
+/*==================[internal data definition]===============================*/
+static uint32_t counts_to_wakeup = 0;
+static uint32_t ciaaPOSIX_counter = 0;
+static uint32_t ciaaPOSIX_sleeps[TASKS_COUNT]={0};
+
+/*==================[external data definition]===============================*/
+
+/*==================[internal functions definition]==========================*/
+static void ciaaPOSIX_sleepAlgorithm(uint32_t toSleep)
+{
+   TaskType taskID;
+   uint32_t aux_counter = getSleepingCounts(ciaaPOSIX_counter);
+
+   /* Are there sleeping tasks? */
+   if(isTaskSleeping(ciaaPOSIX_counter))
+   {
+      if((counts_to_wakeup - aux_counter) > toSleep)
+      {
+         counts_to_wakeup = toSleep + aux_counter;
+
+         /* Remove status bit if any */
+         resetSleepingState(counts_to_wakeup);
+      }
+   }
+   /* So, there is a sleeping task now */
+   else
+   {
+      setSleepingState(ciaaPOSIX_counter);
+
+      counts_to_wakeup = toSleep + aux_counter;
+
+      /* Remove status bit if any */
+      resetSleepingState(counts_to_wakeup);
+   }
+
+   /* Get task id */
+   GetTaskID(&taskID);
+
+   /* Store counts */
+   ciaaPOSIX_sleeps[taskID] = (toSleep + aux_counter);
+
+   /* Set sleeping task state for this task */
+   setSleepingState(ciaaPOSIX_sleeps[taskID]);
+
+   /* wait for the posix event */
+   WaitEvent(POSIXE);
+   ClearEvent(POSIXE);
+}
+
+/*==================[external functions definition]==========================*/
+extern uint32_t ciaaPOSIX_sleep(uint32_t seconds)
+{
+   uint32_t toSleep;
+
+   /* ensure that the seconds can be stored in 10ms counts */
+   ciaaPOSIX_assert((uint32_t)MAX_SECONDS > seconds);
+
+   /* store the sleep time in 10ms counts */
+   toSleep = seconds * SLEEP_TIME_TO_COUNTS;
+
+   /* sleep time processing*/
+   ciaaPOSIX_sleepAlgorithm(toSleep);
+
+   return 0;
+}
+
+extern int32_t ciaaPOSIX_usleep(useconds_t useconds)
+{
+   int32_t ret = 0;
+   uint32_t toSleep;
+
+#if (CIAAPOSIX_DEBUG == CIAAPOSIX_ENABLE)
+   if (1000000 >= useconds)
+   {
+      ret = -1;
+   }
+   else
+#endif
+   {
+      if(MAX_USECONDS > useconds)
+      {
+         /* calculate how many main function cycles shall be sleep */
+         toSleep = (useconds + (CIAAPOSIX_MAINFUNCTION_PERIODUS-1))
+         / CIAAPOSIX_MAINFUNCTION_PERIODUS;
+
+         /* sleep time processing */
+         ciaaPOSIX_sleepAlgorithm(toSleep);
+      }
+      else
+      {
+         ret = -1;
+      }
+   }
+
+   return ret;
+}
+
+extern void ciaaPOSIX_sleepMainFunction(void)
+{
+   uint32_t taskID;
+   uint32_t aux_counter;
+   uint32_t aux_task_counter = 0;
+   uint32_t aux_distance = MAX_COUNTS;
+
+   /* Are there sleeping tasks? */
+   if(isTaskSleeping(ciaaPOSIX_counter))
+   {
+      /* Increment counter */
+      ciaaPOSIX_counter++;
+      /* Set sleeping state bit if there was overflow */
+      setSleepingState(ciaaPOSIX_counter);

```

Comment:
Y si hacemos un bitfield de 1 bit y 31 bits para evitar todo el problema,
entonces además queda mucho más  prolijo? que te parece?

typdef struct {
uint32_t isCouting : 1;
uint32_t counter : 31;
}

no quedaria todo el código más prolijo?

2016-06-12 16:46 GMT+02:00 diegoVom <notifications@github.com>:

> In modules/posix/src/ciaaPOSIX_unistd.c
> <https://github.com/ciaa/Firmware/pull/402#discussion_r66721790>:
>
> > +}
> > +
> > +extern void ciaaPOSIX_sleepMainFunction(void)
> > +{
> > +   uint32_t taskID;
> > +   uint32_t aux_counter;
> > +   uint32_t aux_task_counter = 0;
> > +   uint32_t aux_distance = MAX_COUNTS;
> > +
> > +   /* Are there sleeping tasks? */
> > +   if(isTaskSleeping(ciaaPOSIX_counter))
> > +   {
> > +      /* Increment counter */
> > +      ciaaPOSIX_counter++;
> > +      /* Set sleeping state bit if there was overflow */
> > +      setSleepingState(ciaaPOSIX_counter);
>
> En mi opinion si podria pasar. El bit MSB de ciaaPOSIX_counter esta en 1´
> indicando que hay tareas dormidas, cuando el resto de los bits este en1´
> por incremento del contador la siguiente cuenta será todos en cero
> inclusive el `1´ que indicaba que habia tareas dormidas.
> Propongo encapsular a ciaaPOSIX_counter++; y a
> ciaaLibs_setBit(ciaaPOSIX_counter, 1<<31); en una macro por ejemplo
> ciaaPOSIX_increaseCounter; o algo así. ¿Que opinas?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ciaa/Firmware/pull/402/files/b789d8a5dee28b62b9ed985f82f83f91acd70045#r66721790>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AG7kmysKa8AVB4LksVwE9CSo4IUOgvDTks5qLBvegaJpZM4IzlX2>
> .
>